### PR TITLE
Introduce versioning for proofs

### PR DESCRIPTION
**Pull request by jan-ferdinand**
The explicit version simplifies upgrading proofs for users of Triton VM.

Supersedes #321.

### CODE REVIEW

File: triton-vm/src/proof.rs [hunk 1]
```diff
@@ -78,18 +114,25 @@ impl Claim {
         self.output = output;
         self
     }
+
+    #[must_use]
+    pub fn about_version(mut self, version: u32) -> Self {
```
Comment (aszepieniec) [inline triton-vm/src/proof.rs:119]: Brilliant name :D

File: triton-vm/src/proof.rs [hunk 2]
```diff
@@ -148,4 +191,39 @@ mod tests {
     ) {
         let _proof = Proof::decode(&proof_data);
     }
+
+    #[test]
+    fn current_proof_version_is_still_current() {
+        // todo: Once the prover can be de-randomized, change this test. In particular:
+        //  Seed prover, generate proof, hash the proof, compare to hardcoded digest.
+        //  Remove the proof stored on disk, and remove dependency `bincode`.
```
Comment (aszepieniec) [inline triton-vm/src/proof.rs:199]: I would suggest also referring to the Github issue, and opening one first.

File: triton-vm/src/proof.rs [hunk 3]
```diff
@@ -148,4 +191,39 @@ mod tests {
...
+
+        fn generate_proof_file(program: Program, claim: Claim) {
+            let input = claim.input.clone().into();
+            let non_determinism = NonDeterminism::default();
+            let (aet, _) = VM::trace_execution(&program, input, non_determinism).unwrap();
+            let proof = Stark::default().prove(&claim, &aet).unwrap();
+            let proof = bincode::serialize(&proof).unwrap();
+            fs::create_dir_all("./test_data/").unwrap();
+            fs::write("./test_data/current_version_is_current.proof.new", proof).unwrap();
+            eprintln!("New proof generated. Rename & commit to accept.");
```
Comment (aszepieniec) [inline triton-vm/src/proof.rs:209]: Rename to what? By dropping ".new"? If so, it can't hurt to specify.

**Review by aszepieniec — APPROVED** [review]
Minor nits. Let's go to Mars!

**Comment by Sword-Smith** [comment]
:tada:

**Pull request merged**